File: Week_02/LCOF40/least_numbers2.go
package main

func getLeastNumbers2(arr []int, k int) []int {
	if k == 0 || len(arr) == 0 {
		return []int{}
	}
	if len(arr) == k {
		return arr
	}
	quickSearch(arr, 0, len(arr)-1, k)
	return arr[:k]
}

func quickSearch(arr []int, left, right, k int) {
	if n := partition(arr, left, right); n > k {
		quickSearch(arr, left, n-1, k)
	} else if n < k {
		quickSearch(arr, n, right, k)
	}
}

func partition(arr []int, left, right int) int {
	pivot := arr[right]
	i, j := left-1, left
	for j < right {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
		j++
	}
	i++
	arr[right], arr[i] = arr[i], arr[right]
	return i
}

File: Week_02/LCOF40/least_numbers2.py
from typing import List


class Solution:
    def getLeastNumbers(self, arr: List[int], k: int) -> List[int]:
        if k == 0:
            return []
        n = len(arr)
        if n == 0 or n == k:
            return arr
        self.quick_search(arr, 0, n - 1, k)
        return arr[:k]

    def quick_search(self, arr: List[int], left: int, right: int, k: int):
        n = self.partition(arr, left, right)
        if n > k:
            return self.quick_search(arr, left, n - 1, k)
        elif n < k:
            return self.quick_search(arr, n+1, right, k)

    def partition(self, arr: List[int], left: int, right: int) -> int:
        pivot = arr[right]
        i = left - 1

        for j in range(left, right):
            if arr[j] < pivot:
                i += 1
                arr[i], arr[j] = arr[j], arr[i]
        i += 1
        arr[i], arr[right] = arr[right], arr[i]
        return i

    def test(self):

        for arr, k, target in [
            ([1, 4, 3, 4], 3, [1, 3, 4]),
            ([1, 2, 3, 4], 1, [1]),
            ([1, 2, 3, 4], 1, [1]),
            ([1, 4, 3, 2], 2, [1, 2]),
            ([1, 2, 3, 4], 3, [1, 2, 3]),
            ([1, 2, 3, 4], 4, [1, 2, 3, 4]),
        ]:
            print("nums:", arr)
            ans = sorted(self.getLeastNumbers(arr, k))
            assert ans == target, f"ans: {ans}, target: {target}"
        print("well done")


if __name__ == "__main__":
    Solution().test()
